src/StreamJsonRpc/WebSocketMessageHandler.cs

```diff
@@ -0,0 +1,137 @@
+    /// that uses <see cref="System.Net.WebSockets.WebSocket"/> as the transport.
+    /// </summary>
+    public class WebSocketMessageHandler : DelimitedMessageHandler
+    {
+        private readonly ArraySegment<byte> readBuffer;
+        private readonly byte[] writeBuffer;
+        private Decoder readDecoder;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="WebSocketMessageHandler"/> class.
+        /// </summary>
+        /// <param name="webSocket">The <see cref="System.Net.WebSockets.WebSocket"/> used to communicate.</param>
+        /// <param name="bufferSize">
+        /// The size of the buffer to use for reading JSON-RPC messages.
+        /// Messages which exceed this size will be handled properly but may require multiple I/O operations.
+        /// </param>
+        public WebSocketMessageHandler(WebSocket webSocket, int bufferSize = 4096)
+            : base(Stream.Null, Stream.Null, Encoding.UTF8)
+        {
+            Requires.NotNull(webSocket, nameof(webSocket));
+            Requires.Range(bufferSize > 0, nameof(bufferSize));
+
+            this.WebSocket = webSocket;
+            this.readBuffer = new ArraySegment<byte>(new byte[bufferSize]);
+            this.writeBuffer = new byte[bufferSize];
+        }
+
+        /// <summary>
+        /// Gets the <see cref="System.Net.WebSockets.WebSocket"/> used to communicate.
+        /// </summary>
+        public WebSocket WebSocket { get; }
+
+        /// <inheritdoc />
+        protected async override Task<string> ReadCoreAsync(CancellationToken cancellationToken)
+        {
+            WebSocketReceiveResult result = await this.WebSocket.ReceiveAsync(this.readBuffer, cancellationToken).ConfigureAwait(false);
+            if (result.EndOfMessage)
+            {
+                // fast path: the entire message fit within the buffer.
+                return this.Encoding.GetString(this.readBuffer.Array, 0, result.Count);
+            }
+            else
+            {
+                // The message exceeds the size of our buffer.
+                if (this.readDecoder == null)
+                {
+                    this.readDecoder = this.Encoding.GetDecoder();
+                }
+
+                var jsonBuilder = new StringBuilder();
+                char[] decodedChars = new char[this.readBuffer.Array.Length];
+                void DecodeInput()
+                {
+                    int decodedCharsCount = this.readDecoder.GetChars(this.readBuffer.Array, 0, result.Count, decodedChars, 0, result.EndOfMessage);
+                    jsonBuilder.Append(decodedChars, 0, decodedCharsCount);
+                }
+
+                DecodeInput();
+                while (!result.EndOfMessage)
+                {
+                    result = await this.WebSocket.ReceiveAsync(this.readBuffer, cancellationToken).ConfigureAwait(false);
+                    DecodeInput();
+                }
```

Comment:
why not just read all bytes[] and then do single decoding?

Reply:
Actually your simpler approach is buggy, since it will fail if the incoming message exceeds `this.bufferSize`, since you never grow your buffer.

Reply:
Once you fix that to grow your buffer, you find that you're repeatedly allocating increasingly large buffers and copying the last one into it until you finally find the end size. To avoid that (since that entails GC pressure and a lot of lost processor time), I prefer to reuse the original buffer, and incrementally decode as we go. It should still be super-fast in the normal case, and in the large message case it should keep the throughput high better than just growing the buffer and copying each time.

Reply:
yeah, I realized that after talking to you.  This uses 1 buffer, but increases size of it if need be:
 public class WebSocketMessageHandler : DelimitedMessageHandler, IDisposable
    {
        private int bufferSize;
        private byte[] readBuffer;

        public WebSocketMessageHandler(WebSocket webSocket, int bufferSize = 32768) : base(Stream.Null, Stream.Null, Encoding.UTF8)
        {
            this.bufferSize = bufferSize;
            this.WebSocket = webSocket;
            readBuffer = new byte[bufferSize];
        }

        public WebSocket WebSocket { get; set; }

        protected async override Task<string> ReadCoreAsync(CancellationToken cancellationToken)
        {
            WebSocketReceiveResult result = null;
            int nBytes = 0;
            do
            {
                result = await WebSocket.ReceiveAsync(new ArraySegment<byte>(this.readBuffer, nBytes, this.readBuffer.Length - nBytes), cancellationToken).ConfigureAwait(false);
                nBytes += result.Count;

                // if next read will run out of space (typically reads are 4k reads, but this will be reactive to whatever size we are getting
                if (nBytes > (readBuffer.Length - result.Count))
                {
                    // on rare occasion of big message permentaly auto-grow our buffer size to handle big message size
                    Array.Resize<byte>(ref this.readBuffer, this.readBuffer.Length + this.bufferSize);
                }
            } while (!result.EndOfMessage);
            return this.Encoding.GetString(this.readBuffer, 0, nBytes);
        }

        protected async override Task WriteCoreAsync(string content, Encoding contentEncoding, CancellationToken cancellationToken)
        {
            var buffer = contentEncoding.GetBytes(content);
            await this.WebSocket.SendAsync(new ArraySegment<byte>(buffer), WebSocketMessageType.Text, true, cancellationToken)
                .ConfigureAwait(false);
        }

        protected override void Dispose(bool disposing)
        {
            if (disposing)
            {
                this.WebSocket.Dispose();
            }

            base.Dispose(disposing);
        }
    }

---
In reply to: [160270306](https://github.com/Microsoft/vs-streamjsonrpc/pull/81#discussion_r160270306) [](ancestors = 160270306)

Reply:
Yup: As I said, increasing the buffer size incrementally as you discover the message gets larger is an option. And yes, as you say we could "save" that larger buffer as a field. And yes, that risks one large message keeping a lot of memory allocated ever-after. That's all risk and higher runtime cost than is required if you are willing to perform the work a block at a time. I've written a lot of buffer block style code and I'm quite comfortable with writing and testing such code, so I prefer to do so in the most efficient way possible so I'll never get a bug claiming the library is the cause of their perf problems. :)